### DISCUSSION TITLE
[SDK] Calculation of the required gas `estimateGas`

### DISCUSSION THREAD

**Discussion by RomanYarinski** [zkSync-Community-Hub/zksync-developers#370] [SDKs] [answered]
### Environment

Mainnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/139853615/a338520e-8743-4d77-8516-a32cd66b436f)

There is a dependence of transaction cost estimation on the gas limit.
It turns out that if you use the standard gas estimate, the user will see highly inflated values of transaction cost.
But underestimating the gas limit can be dangerous, because the gas in the zkSync network is not static and can change due to network or L1 load. And if we set the gas limit too low, then the user will not be able to complete the transaction and pay the fee for nothing.

For this reason, there are questions:

1. How can we estimate the gas more accurately so that the user can see a more accurate estimate in Metamask?
2. Can we safely underestimate the resulting estimate to show the user a closer approximation of transaction costs. If we can, by what percentage? Since a transaction that by standard methods is valued at 4.2 million gas actually runs at a gas limit of 200,000.
3. Are there any practices to correct this?
4. Also wondering why the final transaction cost to the user will eat into what gas limit to set? For example, I set the gas limit as estimateGasValue * 0.7 and everything works fine and the transaction is much cheaper for the user after the refund compared to just estimateGasValue. In general, is this logic acceptable, how risky is it and what reduction coefficient can be used?

### Expected Behavior

Estimation of gas showing a close estimate of gas to the actual value so that a user in metamask doesn't see $13 instead of 5 cents of the real value.

### Code Example

With zkSync SDK
``` JavaScript
  const provider = new Provider(chain.rpcUrls.default.http[0])

  const data = encodeFunctionData({
    abi,
    functionName,
    args,
  })

  const zkSyncGas = await provider.estimateGas({
    to: toAddress,
    data,
    value: 0n,
  })
```

with Viem
``` JavaScript
const gas = await publicClient.estimateContractGas({
    address: toAddress,
    abi,
    functionName,
    args,
    account,
  })
```

Equally inaccurate

**Comment by knowalosbra** [comment]
Estimating gas accurately in zkSync transactions poses a challenge due to the dynamic nature of gas costs and limitations in estimating gas usage. Here are some suggestions to address this issue:

Dynamic Gas Estimation: Develop algorithms or models that dynamically estimate gas based on recent network conditions, such as recent gas prices and congestion levels. This can provide a more accurate estimation compared to static gas limits.

Conservative Gas Estimation: Consider adding a safety margin to the gas estimate to account for potential fluctuations in gas costs. While this may result in overestimation in some cases, it ensures that transactions are less likely to fail due to insufficient gas.

User-Friendly Gas Display: Provide users with clear and transparent information about gas estimates and costs in the user interface. This can help manage user expectations and prevent surprises when executing transactions.

Experimentation and Optimization: Continuously monitor gas usage and transaction outcomes to refine gas estimation strategies over time. Experiment with different estimation algorithms and parameters to find the most accurate and cost-effective approach.

Collaboration and Knowledge Sharing: Engage with the zkSync community and other developers to share insights, best practices, and lessons learned in gas estimation. Collaboration can lead to improved techniques and tools for estimating gas in zkSync transactions.

**Reply by RomanYarinski** [reply]
Thanks for the reply, are there any examples of gas estimation algorithms that have proven to work well?

Maybe there are clearer tips from the zkSync team on how to deal with this problem?

Possibly there is some other discussion of this issue that I can read?

**Comment by kaycogrisitbran** [comment]
1. To estimate gas more accurately, consider implementing a custom gas estimation mechanism based on historical data or network conditions.
2. It's generally safe to underestimate gas, but the degree of underestimation depends on the specific use case and network conditions. A conservative approach might involve underestimating by a certain percentage, such as 10-20%, but this should be tested and adjusted accordingly.
3. Best practices include monitoring network gas usage, implementing dynamic gas estimation algorithms, and providing users with transparent fee estimates.
4. Adjusting the gas limit based on estimated gas value can be acceptable, but it's important to consider the risk of transactions failing due to insufficient gas. A reduction coefficient can vary depending on factors such as network congestion and transaction complexity.

**Comment by camicagochristie** [comment]
Accurately estimating gas costs in zkSync transactions is crucial for user experience. Consider adjusting gas estimation algorithms to provide more precise estimates without overestimating. Safely underestimating gas limits can be acceptable, but the percentage should be carefully determined based on network conditions. Implementing practices like setting gas limits as a percentage of estimated gas value and adjusting based on transaction outcomes can help optimize gas usage and improve user cost estimation.

**Comment by StanislavBreadless** [comment, marked as answer]
@RomanYarinski Note, that since state diffs are used, the subsequent transactions in a batch can have smaller fee due to the fact that they reused the state storage slots. So depending on the sort of the contract you are calling you can get very different cost of the first and the second call, but when estimating gas we never know whether your transaction will come as one of the first ones in a batch (and will have to pay the full price) or it will continue receiving the benefits of the previous transactions in the current batch that have already paid for the state changes.

Also, the calculations in the table seem to be a bit off. On the 23rd of February the gas price on zksync was around 0.1 gwei, in other words 4200000 gas would probably cost around `4200000 * 0.1 * 3500 / 1000000000 = $1.47` (I took the numbers from my head, but most certainly the price wasn't much different from this number), which is still off from the final gas used, but not by the large margin that is provided in the table
